request-deferrer/service-worker.js

```diff
@@ -0,0 +1,182 @@
+// get the id for the new quotations. No problem, just say we accept the
+// creation and we will process it later, as soon as we recover connectivity.
+worker.post(root + 'api/quotations?*', tryOrFallback(new Response(null, {
+  status: 202
+})));
+
+// Lastly, this a virtual route to force the a sync. It is originated
+// by the client who is listening for the `online` event. This request
+// never reaches the network. It is here to be consistent but it is pure
+// virtual.
+worker.post(root + 'api/sync', function() {
+  return flushQueue().then(function() {
+    return new Response();
+  });
+});
+
+// Start the service worker.
+worker.init();
+
+// By using Mozilla's localforage db wrapper, we count with
+// a fast setup for a versatile key, value database. We use
+// it to store queue of deferred requests.
+
+// Enqueue consists into add to the list a request. Due to the
+// limitations of IndexedDB, Request and Response objects can not
+// be saved so we need an alternative representations. This is
+// why we call to `serialize()`.`
+function enqueue(request) {
+  return serialize(request).then(function(serialized) {
+    localforage.getItem('queue').then(function(queue) {
+      /* eslint no-param-reassign: 0 */
+      queue = queue || [];
+      queue.push(serialized);
+      return localforage.setItem('queue', queue).then(function() {
+        console.log(serialized.method, serialized.url, 'enqueued!');
+      });
+    });
+  });
+}
+
+// Flush is a little more complicated. It consists into get
+// the elements of the queue in order and send each one,
+// keeping track of not yet sent request. Before sending a request
+// we need to recreate it from the alternative representation
+// stored in IndexedDB.
+function flushQueue() {
+  // Get the queue
+  return localforage.getItem('queue').then(function(queue) {
+    /* eslint no-param-reassign: 0 */
+    queue = queue || [];
+
+    // If empty, nothing to do!
+    if (!queue.length) {
+      return Promise.resolve();
+    }
+
+    // Else, send the requests in order...
+    console.log('Sending ', queue.length, ' requests...');
+    return sendInOrder(queue).then(function() {
+      // **Requires error handling**. Actually, this is assuming all the requests
+      // in queue are a success when reaching the Network. So it should empty the
+      // queue step by step, only popping from the queue if the request completes
+      // with success.
+      return localforage.setItem('queue', []);
+    });
+  });
+}
+
+// Send the requests inside the queue in order. Waiting for the current before
+// sending the next one.
+function sendInOrder(requests) {
+  var sending = Promise.resolve();
+  requests.forEach(function(serialized) {
+    console.log('Sending', serialized.method, serialized.url);
+    sending = sending.then(function() {
+      return deserialize(serialized).then(function(request) {
+        fetch(request);
+      });
+    });
+  });
+  return sending;
```

Comment:
Nit: it would be nice to use `reduce` here:
```JavaScript
return requests.reduce(function(prevPromise, serialized) {
    console.log('Sending', serialized.method, serialized.url);
    return prevPromise.then(function() {
      return deserialize(serialized).then(function(request) {
        fetch(request);
      });
    });
}, Promise.resolve());
```

P.S.: To send the requests in order, we should also wait the fetch promise to be resolved, so instead of `fetch(request)` -> `return fetch(request);`